File: builtinsocket.py
#!/usr/bin/env python3

import sys
import socket
import json


def main():
    # Validate minimum arguments: action flag and socket path
    if len(sys.argv) < 3:
        print("0")
        sys.exit(1)

    act = sys.argv[1]
    ipc_sock = sys.argv[2]

    # Read payload data dispatched from Bash herestring (STDIN)
    input_data = ""
    if not sys.stdin.isatty():
        input_data = sys.stdin.read().strip()

    # ESTABLISH CONNECTION TO THE UNIX DOMAIN SOCKET OF MPV
    try:
        s = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
        s.connect(ipc_sock)
    except Exception:
        if act == "--cache-bytes":
            print("0")
        sys.exit(1)

    try:
        # 1. FETCH STANDARD STRING PROPERTIES (Title, Uploader, Time-pos)
        if act == "--get-prop":
            req = json.dumps({"command": ["get_property", input_data]}) + "\n"
            s.sendall(req.encode("utf-8"))
            
            # Retrieve response dataset frames from the stream buffer
            raw = s.recv(4096).decode("utf-8").strip()
            for ln in raw.split("\n"):
                if not ln:
                    continue
                obj = json.loads(ln)
                if "data" in obj and obj["data"] is not None:
                    print(obj["data"])
                    break

        # 2. DISPATCH AUDIO CONTROL COMMAND MATRIX (Seek, Pause, Resume)
        elif act == "--send-cmd":
            args = json.loads(input_data)
            req = json.dumps({"command": args}) + "\n"
            s.sendall(req.encode("utf-8"))

        # 3. PARSE COMPLEX HASH-MAP STRUCTURE FOR DEMUXER CACHE METRICS
        elif act == "--cache-bytes":
            req = json.dumps({"command": ["get_property", "demuxer-cache-state"]}) + "\n"
            s.sendall(req.encode("utf-8"))
            raw = s.recv(4096).decode("utf-8").strip()
            for ln in raw.split("\n"):
                if not ln:
                    continue
                obj = json.loads(ln)
                if "data" in obj and isinstance(obj["data"], dict) and "total-bytes" in obj["data"]:
                    print(obj["data"]["total-bytes"])
                    break
            else:
                print("0")

    except Exception:
        if act == "--cache-bytes":
            print("0")
    finally:
        try:
            s.close()
        except Exception:
            pass

if __name__ == "__main__":
    main()

File: builtinsocket.js
#!/usr/bin/env node

const fs = require("fs");
const net = require("net");

(function () {
    // Validate minimum command-line execution parameters: action and socket path
    if (process.argv.length < 4) {
        console.log("0");
        process.exit(1);
    }

    const act = process.argv[2];
    const ipcSock = process.argv[3];

    // Read payload parameters dispatched from Bash herestring via STDIN (FD 0)
    let inputData = "";
    try {
        if (!process.stdin.isTTY) {
            inputData = fs.readFileSync(0, "utf-8").trim();
        }
    } catch (e) {
        inputData = "";
    }

    // ESTABLISH UNIDIRECTIONAL CONNECTION HOOK TO THE MPV UNIX SOCKET
    const client = net.createConnection(ipcSock, () => {
        // 1. STRIP AND ROUTE VALUE PROPERTIES (Title, Uploader, Progress Time)
        if (act === "--get-prop") {
            client.write(JSON.stringify({ "command": ["get_property", inputData] }) + "\n");
        }
        // 2. DISPATCH INTERACTIVE CONTROL ARRAYS (Seek, Pause, Resume)
        else if (act === "--send-cmd") {
            try {
                const args = JSON.parse(inputData);
                client.write(JSON.stringify({ "command": args }) + "\n", () => {
                    client.destroy();
                    process.exit(0);
                });
            } catch (e) {
                client.destroy();
                process.exit(1);
            }
        }
        // 3. UNROLL ADVANCED HASHED HIERARCHIES FOR DEMUXER STREAM CACHE SIZE
        else if (act === "--cache-bytes") {
            client.write(JSON.stringify({ "command": ["get_property", "demuxer-cache-state"] }) + "\n");
        }
    });

    // HANDLE INCOMING RESPONSE STREAM FROM MPV
    client.on("data", (data) => {
        try {
            const rawStr = data.toString().trim();
            // Isolate the very first JSON line returned by the socket buffer
            const fstLn = rawStr.split("\n")[0];
            if (!fstLn) return;
            const obj = JSON.parse(fstLn);
            if (act === "--get-prop") {
                if (obj.data !== undefined && obj.data !== null) {
                    console.log(obj.data);
                }
                client.destroy();
                process.exit(0);
            } 
            else if (act === "--cache-bytes") {
                if (obj.data && obj.data["total-bytes"] !== undefined) {
                    console.log(obj.data["total-bytes"]);
                } else {
                    console.log(0);
                }
                client.destroy();
                process.exit(0);
            }
        } catch (e) {
            if (act === "--cache-bytes") console.log(0);
            client.destroy();
            process.exit(1);
        }
    });

    // FALLBACK SAFETY INTERCEPTORS TO PREVENT HANGING PROCESSES
    client.on("error", () => {
        if (act === "--cache-bytes") console.log(0);
        client.destroy();
        process.exit(1);
    });

    client.on("end", () => {
        client.destroy();
    });
})();
